File: dotnet/src/Day7/Puzzle.cs
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode.Day7
{
    public static class Puzzle
    {
        public static Program SolvePart1(Program[] programs)
        {
            return programs.Single(program => !IsProgramAboveOtherProgram(program, programs));
        }

        private static bool IsProgramAboveOtherProgram(Program program, IEnumerable<Program> programs) =>
            programs.Any(p => p.ProgramsAbove.Contains(program.Name));
    }
}

File: 2017/go/day7/puzzle.go
package day7

type Program struct {
	Name string
	Weight int
	ProgramsAbove []string
}

func SolvePart1(programs []Program) Program {
	for _, program := range programs {
		if !isProgramAboveOtherProgram(program.Name, programs) {
			return program
		}
	}

	panic("bottom program not found")
}

func isProgramAboveOtherProgram(programName string, programs []Program) bool {
	for _, program:= range programs {
		for _, programAbove := range program.ProgramsAbove {
			if programAbove == programName {
				return true
			}
		}
	}

	return false
}
